source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -244,14 +271,112 @@ bool Filter::parseExtensions(os_fd_t fd) {
     if (bytes_avail == 0) {
       return false;
     }
-    bytes_avail = std::min(size_t(bytes_avail), sizeof(buf_));
+    bytes_avail = std::min(size_t(bytes_avail), buf_size);
     bytes_avail = std::min(size_t(bytes_avail), proxy_protocol_header_.value().extensions_length_);
-    const Api::SysCallSizeResult recv_result = os_syscalls.recv(fd, buf_, bytes_avail, 0);
+    buf += buf_off ? *buf_off : 0;
+    const Api::SysCallSizeResult recv_result = os_syscalls.recv(fd, buf, bytes_avail, 0);
     if (recv_result.rc_ != bytes_avail) {
       throw EnvoyException("failed to read proxy protocol extension");
     }
     proxy_protocol_header_.value().extensions_length_ -= recv_result.rc_;
+
+    if (buf_off) {
+      *buf_off += recv_result.rc_;
+    }
+  }
+
+  return true;
+}
+
+/**
+ * @note  A TLV is arranged in the following format: @n
+ *        struct pp2_tlv {      @n
+            uint8_t type;       @n
+            uint8_t length_hi;  @n
+            uint8_t length_lo;  @n
+            uint8_t value[0];   @n
+          };                    @n
+ *        See https://www.haproxy.org/download/2.1/doc/proxy-protocol.txt for details
+ */
+void Filter::parseTlvs(const std::vector<uint8_t>& tlvs) {
+  size_t size = tlvs.size();
+  uint8_t tlv_type;
```

Comment:
Declare these variables where they are first used, not at top of function.

Reply:
Code updated as suggested.